### ISSUE TITLE
systemd-boot support

### ISSUE THREAD

**Issue by kirbyfan64** [ostreedev/ostree#1719]
From what I can tell, this is basically present (since the loader syntax is identical), except systemd-boot needs the loaders and kernels to be on the ESP (/boot/efi), not where OSTree places them currently (/boot)...

**Comment by cgwalters** [comment]
That sounds likely yes.  I have no opposition to supporting `sd-boot` - the main question is how we detect it (actually the fact that we dynamically detect bootloaders today is a bit of a mess).

**Comment by kirbyfan64** [comment]
Maybe just check for its EFI file? GRUB is a little easily to have automatically installed accidentally forget about it, but systemd-boot is usually manually (and therefore consciously) installed.

**Comment by AdrianVovk** [comment]
I've run into this issue also. I'm trying to develop a distro around OSTree and I'd much rather use systemd-boot than grub for efi

**Comment by AdrianVovk** [comment]
@cgwalters Can the bootloader detection issue be solved with a simple config file? OSTree, when running `ostree admin deploy` can simply look for a file in /usr/etc or whatever and see which bootloader it should use.

Another possibility is to have the installer tell OSTree which bootloader to use. I'd prefer this option, since this gives the installer script the flexibility to decide on the bootloader, not the tree. This way, boot installation/config all happens in one place. Something like: `ostree admin init-boot --sysroot=/mnt --os=foo systemd-boot`

If the config file doesn't exist/command wasn't run, OSTree can (temporarily) fall back to the current behavior

**Comment by cgwalters** [comment]
For people who are building libostree from source (as it sounds like you are?), one thing we could pretty easily do is add a build-time option, like `--with-bootloader=sd-boot` or so.  Would that be sufficient?

My main hesistation on this is where to put the config file, but thinking about it as I'm typing, we could add it as an entry in the repo config?  Something like

```
[ostree-sysroot]
bootloader=<grub|sd-boot>
```

**Comment by AdrianVovk** [comment]
@cgwalters Yes, I am building libostree from source. An option like this would be sufficient. The problem with it is that it will restrict the entire system to a single bootloader. If the user prefers Grub and replaces sd-boot, suddenly OSTree deployments break. You'd need to recompile OSTree to boot with Grub. For minimal distros, though, it would be nice to trim down ostree by restricting bootloader support

Putting it in the repo config would be perfect! This also automatically gives a cli command, since `ostree config` exists. I would suggest maybe naming the option `boot.loader` or `system.bootloader`. `ostree-sysroot` seems unwieldy...

**Comment by kirbyfan64** [comment]
FWIW ostree also does stuff like chmod the kernel file's permissions when it saves it to /boot, but that can't be done with systemd-boot (since it only supports kernels on the ESP, which is fat).

**Comment by AdrianVovk** [comment]
@kirbyfan64 can it not just skip that step? I've been manually copying things from /boot to /boot/efi without issue

**Comment by kirbyfan64** [comment]
It could, I'm just pointing out that it would also need to be changed

**Comment by AdrianVovk** [comment]
Any update on this?

**Comment by AdrianVovk** [comment]
@cgwalters I've been able to mess about with my distro and get some insight about the relationship between `/boot`, the ESP, and systemd

systemd, by default, really wants the ESP to be mounted to either `/boot` directly or `/efi`. If `/efi` exists, systemd will mount the ESP there. Otherwise, it will try to mount it to `/boot`. This is the behaviour of the `systemd-gpt-auto-generator`, which my distro will use to find and mount the rootfs. `bootctl` checks those two mounts points for an ESP also. It's safe to assume that any systemd-related tool will probably assume the EFI exists in either of these two directories (short of looking at /etc/fstab, which I don't intend to use)

I've been able to temporarily get OSTree working like so: I have a `boot.mount` unit which bind-mounts `/sysroot/boot` to `/boot`. OSTree can then operate on this directory normally. I created an `/efi` mount point (and systemd automagically mounts the ESP there) that I can copy files to as necessary. This, of course, is not an atomic layout. If something goes wrong while my distro's scripts are copying from `/boot` to `/efi`, I'm screwed. The system would no longer boot.

OSTree depends on symlinks in `/boot`, so it cannot operate directly on the ESP. If this dependency can be broken, however, it could be possible to work with the ESP in `/boot`. I think this can be done with the `renameat2` syscall to atomically swap two folders, instead of using a symlink and `rename`. [More info](http://man7.org/linux/man-pages/man2/rename.2.html) and [initial patch](https://lwn.net/Articles/569028/). The new flow could go something like this: empty out `/boot/loader.inactive/`, generate the ostree config files, copy in non-ostree config files from `/boot/loader`, `renameat2` to swap the two folders. This would give an atomic swap without relying on symlinks. If I'm not missing something, this would allow OSTree to work directly on the ESP. Steps like setting permissions on the files can be skipped or can error out silently, since vfat doesn't support permissions anyway AFAIK

If, for compatability, you need to keep the old behavior, you can translate all of this new behavior to the `/efi` directory. This new atomic swap would happen in `/efi`, and the old behavior would happen in `/boot`.

What are your thoughts?

**Comment by AdrianVovk** [comment]
PS: It would also be nice is OSTree could format its filenames for [boot counting](https://www.freedesktop.org/software/systemd/man/systemd-boot.html#Boot%20Counting)

**Comment by cgwalters** [comment]
There's extensive prior discussion of EFI and BLS and ostree - see e.g. https://github.com/ostreedev/ostree/issues/717

Let's keep this issue just about having the bootloader be configurable.

**Comment by AdrianVovk** [comment]
@cgwalters Certainly. That issue seems more grub oriented and less EFI but I moved the ESP discussion to #1649. Moreover, isn't the issue of supporting sd-boot really just an issue of supporting the ESP?

I don't see any problems with having a simple repo config option for determining which files to generate on deploy. Maybe you could even support having a list, and if absent the value would be `grub,bls` for backwards compatibility reasons

You could support sd-boot's bootcounting the same way. A config option would determine the number appended to the end of generated boot options

Like so:
```
[boot]
# Use the Boot Loader Standard
loader=bls
# Boot couting (appends +3 to names of generated bls files)
retry=3
```

**Comment by klausenbusk** [comment]
> @cgwalters Can the bootloader detection issue be solved with a simple config file? OSTree, when running `ostree admin deploy` can simply look for a file in /usr/etc or whatever and see which bootloader it should use.

`sd-boot` can be detected by checking the EFI variable `LoaderInfo`, but using a config file is probably better.
```
$ cat /sys/firmware/efi/efivars/LoaderInfo-4a67b082-0a4c-41cf-b6c7-440b29bb8c4f 
systemd-boot 240
```

**Comment by fansari** [comment]
I used systemd-boot ("gummitboot") with "normal" Fedora in the past. As I remember this worked fine - but if you want to use "Secure Boot" you had to use this solution.

https://blog.hansenpartnership.com/linux-foundation-secure-boot-system-released/

As far as I understood secure boot has to been done in a complete different way wiht systemd-boot. You cannot just sign your kernel but have always add a hash signature to the MokList. I had one PC where this did not work. So I don't know whether this is a solution which will be maintained in future.

And if systemd-boot is only usable if you disable Secure Boot then I would stay with grub. But only for this reason - otherwise I would always prefer systemd-boot.

**Comment by AdrianVovk** [comment]
@fansari you should research shim. It allows two different ways to do secure boot: the MokList, basically a list of hashes of EFI binaries that are safe to run, and with signed binaries. I think systemd-boot has little to do with it to be honest

Also, unless I'm missing something this is seemingly unrelated to the issue.

**Comment by bam80** [comment]
Just wanted to clarify the status. Can we have milestone for this or something, to track it better?

**Comment by gdonval** [comment]
> I think this can be done with the renameat2 syscall to atomically swap two folders, instead of using a symlink and rename.

Looks like the prerequisites for #1967, proper atomic `renameat2` on `vfat`, are released (Linux 6.0).

Also, this is not necessary if boot counting is leveraged: there can be always one known-good version present if done correctly. The `install-kernel` mechanics can be used to ensure that it is "done correctly".

Concurrently, unified kernels made their ways into the fray with the exact same boot counting semantics. One way `ostree` could become compatible with `sd-boot` at low programming cost without breaking legacy BIOS would be by generating (and optionally signing) unified kernels in the right place in the `EFI` partition.

**Comment by Apacelus** [comment]
> I was able to switch to systemd-boot on my Fedora Silverblue 37 installation by changing the partition type of `/boot` to Extended Boot Partition from GNOME Disks and then running `bootctl install`. Seems to work fine without much problems.
> 
> EDIT: I had to also install the ext4 EFI file system driver from [efifs](https://github.com/pbatard/efifs) into my EFI partition to get sd-boot to recognize my /boot partition.

@godvino can you please explain what exact files you got from efifs(they only have ext2 files?) and where exactly you put them?

**Comment by fwilhe** [comment]
Hey everyone, I'm looking to build ostree-based [Garden Linux](https://gardenlinux.io/) system, and for GL we absolutely require systemd-boot instead of grub. In this POC stage, building ostree myself (also from a non-main branch) would be fine, but in the long run it would be great if this was a standard feature of ostree.

Can someone give me some pointers on the current status of this? I'm not yet familiar with ostree internals, but I'd be okay to spend time looking into this.

**Comment by ericcurtin** [comment]
@fwilhe if Garden Linux does non-UKI bls-based booting, the contributions required to get sd-boot working should be minimal... UKI support is in progress...

**Comment by fwilhe** [comment]
Hi @ericcurtin,

thanks for your hint, I did make quite some progress meanwhile and I got a booting system using systemd-boot. I did not expect this to work and it might actually not be working properly, at least that's my suspicion for the cause of issue #3022.

Assuming non-UKI setup (which is good enough for now), can you give me some pointers of what kind of contributions you mean? I did look into the bootloader specific code of ostree, but I'm not sure what systemd-boot specific code would need to do.

Best,
Florian

**Comment by ericcurtin** [comment]
@fwilhe I don't actually know what specifically needs to change, just gotta keep building ostree based images with sd-boot and see what isn't working I guess...

**Comment by GrabbenD** [comment]
> I recommend looking at the ostree osbuild python script stages to get an idea of how to compose an ostree-based OS.

Where can this be found @ericcurtin?

**Comment by ericcurtin** [comment]
The ostree python scripts here:

https://github.com/osbuild/osbuild/tree/main/stages

that's how we build ostree based OS's, the log I posted above describes the order in which they are called.

**Comment by ericcurtin** [comment]
There's probably grub assumptions you are gonna come across that need to be sorted out.

**Comment by ericcurtin** [comment]
This would need to be completed also, not sure if @valentindavid wants to complete or if he wants to hand it off to someone else.

**Comment by fwilhe** [comment]
> @fwilhe I don't actually know what specifically needs to change, just gotta keep building ostree based images with sd-boot and see what isn't working I guess...

Are there any existing tests/PoCs/example for that?

Not sure what it's worth, but I've put my debian/sd-boot based experiment [here](https://github.com/fwilhe/ostree-debian-builder). It's booting, but it is not really functioning.

**Comment by p1u3o** [comment]
> > I was able to switch to systemd-boot on my Fedora Silverblue 37 installation by changing the partition type of `/boot` to Extended Boot Partition from GNOME Disks and then running `bootctl install`. Seems to work fine without much problems.
> > EDIT: I had to also install the ext4 EFI file system driver from [efifs](https://github.com/pbatard/efifs) into my EFI partition to get sd-boot to recognize my /boot partition.
> 
> @godvino can you please explain what exact files you got from efifs(they only have ext2 files?) and where exactly you put them?

I haven't seen anyone reply to this. So I will document how I got systemd-boot working on Silverblue 39. So far, everything works. Make sure **secure boot is disabled**.

```
# # Set your /boot partition to "Extended Boot Partition" in GNOME Disks (see https://i.imgur.com/Tr4Z41o.png)
# rpm-ostree install systemd-boot-unsigned
# rpm-ostree apply-live
# mkdir -p /boot/efi/EFI/systemd/drivers
# cd /boot/efi/EFI/systemd/drivers
# wget https://github.com/pbatard/efifs/releases/download/v1.9/ext2_x64.efi
# bootctl install
```

This has the benefit of also not putting kernel images on the efi partition, which is usually smaller than /boot. Grub will continue to work as well.

**Comment by bam80** [comment]
Any update?

**Comment by bam80** [comment]
> I will try to remember it. I will see if I can find time to rebase and address the directory version concern.

@valentindavid ping

**Comment by bam80** [comment]
> `# # Set your /boot partition to "Extended Boot Partition" in GNOME Disks`

After transitioning to a separate /boot partition I started to see this error:
```sh
$ bootctl 
Failed to open '/boot//loader/entries': Remote address changed
...
```
this is with systemd-boot or grub, regardless.
The system boots, though.

If I return to my previous /boot placement (on btrfs subvolume), I see no such error.

I have no idea where this error comes from.
Thoughts?

**Comment by gdonval** [comment]
> Make sure **secure boot is disabled**.

I am personally not keen on adding attack surface to `sd-boot` by disabling a crucial component of verified (to-the-admin) boot and add complex code in the form of a `ext2` driver.

I've had success mounting the ESP on `/efi` and use kernel-install to automatically create (signed) UKIs from ostree-controlled `/boot` binaries. 

The problem is that it's very clunky at the moment and the boot-blessing mechanism is different from the bootloader specs'.

Actually, if fcos could create UKIs and use grub's own (undocumented) support for the bootloader spec, switching bootloaders would become nearly trivial.

**Comment by bam80** [comment]
> > `# # Set your /boot partition to "Extended Boot Partition" in GNOME Disks`
> 
> After transitioning to a separate /boot partition I started to see this error:
> 
> ```shell
> $ bootctl 
> Failed to open '/boot//loader/entries': Remote address changed
> ...
> ```
> 
> this is with systemd-boot or grub, regardless. The system boots, though.

Any advice on this?
I upgraded from SB 40 to SB 41 and now my system toggles the boot deployment on every reboot between SB 41 and SB39.
Have no idea how to debug that, but maybe that error is somehow related..

**Comment by bam80** [comment]
> After transitioning to a separate /boot partition I started to see this error:
> 
> ```shell
> $ bootctl 
> Failed to open '/boot//loader/entries': Remote address changed
> ...
> ```

https://github.com/systemd/systemd/issues/35293

However, it might be an issue with SB.

**Comment by igoropaniuk** [comment]
@AdrianVovk  Regarding boot counting support, I've created a PR for this https://github.com/ostreedev/ostree/pull/3310

**Comment by bam80** [comment]
I encapsulated slightly modified sd-boot installation commands suggested by @p1u3o to a systemd unit.

It disables itself after successful sd-boot install, and runs on (the first) boot, or manually.
The unit can be useful for system installation scripts, I personally run it from `%post` Kickstart section.

It also can be started again later to update `EfiFs` driver (and `sd-boot` itself, too):
```systemd.service
sudo sh -c 'cat > /etc/systemd/system/firstboot-sdboot.service' <<-EOF
	[Unit]
	Description=Install sd-boot

	#Before=firstboot-rpm-ostree.service
	After=network-online.target
	Wants=network-online.target

	[Service]
	ExecStart=rpm-ostree install --apply-live --idempotent systemd-boot-unsigned
	#ExecStart=wget --directory-prefix /boot/efi/EFI/systemd/drivers https://github.com/pbatard/EfiFs/releases/latest/download/ext2_x64.efi
	ExecStart=curl --location --remote-name https://github.com/pbatard/EfiFs/releases/latest/download/ext2_x64.efi --output-dir /boot/efi/EFI/systemd/drivers --create-dirs
	ExecStart=bootctl install

	ExecStop=systemctl disable %N
	#ExecStop=rm -v %y

	StandardOutput=journal+console
	Type=oneshot

	[Install]
	WantedBy=multi-user.target
EOF

# if you want it run (once) on the next boot
systemctl enable firstboot-sdboot
systemctl status firstboot-sdboot

# if you want to start it manually
systemctl start firstboot-sdboot
systemctl status firstboot-sdboot
```
Suggestions to improve are welcome.

**Comment by p1u3o** [comment]
I'd suggest doing some kind of checksum check on pulling the EFI driver there, if that repo is compromised you'll be giving the door for malware into the EFI firmware.

**Comment by gitnohubz** [comment]
I hope kinoite 43 whould get systemdboot support

**Comment by ta-vroom** [comment]
I was also able to get sd-boot installed and working using @p1u3o's [method](https://github.com/ostreedev/ostree/issues/1719#issuecomment-1749394763). This feels rather hacky given that sd-boot does not support symlinks or other filesystems natively. If we have to use the efi partition for writing entries, is there a way to preserve atomicity? 

What about writing a new "generation".conf (like NixOS) and increment it so that sd-boot always chooses the latest version (which sd-boot does by default)?

ostree0.conf -----> previous boot
ostree1.conf -----> current boot
rpm-ostree operation -----> ostree2.conf 
if write fails == error
if write succeeds == delete ostree0.conf

future rpm-ostree operations would write to ostree3.conf, ostree4.conf, and so on...
